### PR TITLE
Added reservation action

### PR DESCRIPTION
**Pull request by abdullokhtolibjonov**

### CODE REVIEW

File: Parking-Zone/Areas/User/Controllers/ReservationController.cs [hunk 1]
```diff
@@ -41,5 +42,34 @@ public IActionResult FreeSlots(FreeSlotsVM freeSlotsVM)
...
+
+        [HttpPost]
+        public IActionResult Reserve(ReserveVM reserveVM)
+        {
+            var slot = _slotService.GetById(reserveVM.SlotId);
+
+            if (ModelState.IsValid)
+            {
+                if (_slotService.IsSlotFree(slot, DateTime.Parse(reserveVM.StartTime), reserveVM.Duration))
+                {
```
Comment (DilshodMQ) [inline Parking-Zone/Areas/User/Controllers/ReservationController.cs]: should be added model error if Slot is not free and show this error in the view

File: Parking-Zone/Areas/User/Controllers/ReservationController.cs [hunk 2]
```diff
@@ -41,5 +42,34 @@ public IActionResult FreeSlots(FreeSlotsVM freeSlotsVM)
...
+                    var reservation = reserveVM.MapToModel();
+                    _reservationService.Insert(reservation);
```
Comment (DilshodMQ) [inline Parking-Zone/Areas/User/Controllers/ReservationController.cs]: Who has create this Reservation?  Add FK to AppUser from Reservation Model and attach the current User

File: Parking-Zone/Areas/User/Views/Reservation/Reserve.cshtml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    <h6>Slot Number: @Model.SlotNumber</h6>
+    <hr />
+    <div class="col-md-4">
+        <form asp-action="Reserve">
+            <div asp-validation-summary="ModelOnly" class="text-danger"></div>
+
+            <input type="hidden" value="@Model.SlotId" asp-for="SlotId" />
+            <input type="hidden" value="@Model.ZoneName" asp-for="ZoneName" />
+            <input type="hidden" value="@Model.ZoneAddress" asp-for="ZoneAddress" />
+            <input type="hidden" value="@Model.SlotNumber" asp-for="SlotNumber" />
```
Comment (DilshodMQ) [inline Parking-Zone/Areas/User/Views/Reservation/Reserve.cshtml]: Why you need the last three input? only slotID is enaugh

File: Parking-Zone/Areas/User/Controllers/ReservationController.cs [hunk 4]
```diff
@@ -41,5 +42,34 @@ public IActionResult FreeSlots(FreeSlotsVM freeSlotsVM)
 
             return View(freeSlotsVM);
         }
+
+        public IActionResult Reserve(Guid SlotId, string startTime, int duration)
+        {
+            var slot = _slotService.GetById(SlotId);
+
```
Comment (DilshodMQ) [inline Parking-Zone/Areas/User/Controllers/ReservationController.cs:56]: what if slot ==null (also test for this case)

File: Tests/Controllers/ReservationControllerTests.cs [hunk 5]
```diff
@@ -109,5 +127,71 @@ public void GivenReservationVMModel_WhenPostFreeSlotsCalled_ThenServiceCalledOnc
...
+            //Arrange
+            var testStartTime = "2024-01-27 18:00:00";
+            var expectedReserveVM = new ReserveVM(_testSlot, testStartTime, 2);
+
+            mockSlotService
+                .Setup(service => service.GetById(_testSlotId))
+                .Returns(_testSlot);
+
+            //Act
+            var result = controller.Reserve(_testSlotId, testStartTime, 2);
```
Comment (DilshodMQ) [inline Tests/Controllers/ReservationControllerTests.cs:159]: create var for duration and use it

**Review by DilshodMQ — CHANGES_REQUESTED** [review]

File: Parking-Zone/ViewModels/Reservation/ReserveVM.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        public ReserveVM(Models.ParkingSlot slot, string startTime, int duration)
+        {
+            StartTime = startTime;
+            Duration = duration;
+            SlotId = slot.Id;
+            ZoneAddress = slot.ParkingZone.Address;
+            ZoneName = slot.ParkingZone.Name;
+            SlotNumber = slot.Number;
+        }
+        public Guid SlotId { get; set; }
```
Comment (napster1991) [inline Parking-Zone/ViewModels/Reservation/ReserveVM.cs:23]: Also required

File: Parking-Zone/Areas/User/Controllers/ReservationController.cs [hunk 7]
```diff
@@ -41,5 +51,49 @@ public IActionResult FreeSlots(FreeSlotsVM freeSlotsVM)
...
+                return NotFound();
+
+            reserveVM.SlotNumber = slot.Number;
+            reserveVM.ZoneAddress = slot.ParkingZone.Address;
+            reserveVM.ZoneName = slot.ParkingZone.Name;
+
+            if (_slotService.IsSlotFree(slot, DateTime.Parse(reserveVM.StartTime), reserveVM.Duration))
+            {
+                var reservation = reserveVM.MapToModel();
+                reservation.AppUserId = _userManager.GetUserId(User);
```
Comment (napster1991) [inline Parking-Zone/Areas/User/Controllers/ReservationController.cs]: Fetch userId from context Identity. Then you can remove userManager from controller.

File: Parking-Zone/Areas/User/Views/Reservation/Reserve.cshtml [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+<div class="row">
+    <h2>@Model.ZoneName , @Model.ZoneAddress</h2>
+    <h6>Slot Number: @Model.SlotNumber</h6>
+    <hr />
+    <div class="col-md-4">
+        <form asp-action="Reserve">
+            <div asp-validation-summary="ModelOnly" class="text-danger"></div>
+
+            <input type="hidden" value="@Model.SlotId" asp-for="SlotId" />
```
Comment (napster1991) [inline Parking-Zone/Areas/User/Views/Reservation/Reserve.cshtml]: I guess, no need in setting the default value.

File: Tests/Controllers/ReservationControllerTests.cs [hunk 9]
```diff
@@ -109,5 +131,76 @@ public void GivenReservationVMModel_WhenPostFreeSlotsCalled_ThenServiceCalledOnc
             Assert.Equal(JsonSerializer.Serialize(expectedSlotsVM), JsonSerializer.Serialize((result as ViewResult).Model));
             mockSlotService.Verify(service => service.GetFreeByZoneIdAndTimePeriod(_testZoneId, testStartTime, 3), Times.Once);
         }
+
```
Comment (napster1991) [inline Tests/Controllers/ReservationControllerTests.cs:144]: Please, add regions.

File: Tests/Controllers/ReservationControllerTests.cs [hunk 10]
```diff
@@ -109,5 +131,76 @@ public void GivenReservationVMModel_WhenPostFreeSlotsCalled_ThenServiceCalledOnc
...
+            //Act
+            var result = controller.Reserve(_testSlotId, testStartTime, 2);
+
+            //Assert
+            Assert.IsType<ViewResult>(result);
+            Assert.Equal(JsonSerializer.Serialize(expectedReserveVM), JsonSerializer.Serialize((result as ViewResult).Model));
+            mockSlotService.Verify(service => service.GetById(_testSlotId), Times.Once);
+        }
+
```
Comment (napster1991) [inline Tests/Controllers/ReservationControllerTests.cs:166]: Add test case, about which Dilshod mentioned

File: Tests/Controllers/ReservationControllerTests.cs [hunk 11]
```diff
@@ -109,5 +131,76 @@ public void GivenReservationVMModel_WhenPostFreeSlotsCalled_ThenServiceCalledOnc
...
+            controller.ModelState.AddModelError("VehicleNumber", "Vehicle Number is Required");
+
+            //Act
+            var result = controller.Reserve(reserveVM);
+
+            //Assert
+            Assert.False(controller.ModelState.IsValid);
+            Assert.IsType<ViewResult>(result);
+            Assert.Equal(JsonSerializer.Serialize(reserveVM), JsonSerializer.Serialize((result as ViewResult).Model));
+        }
```
Comment (napster1991) [inline Tests/Controllers/ReservationControllerTests.cs:235]: Add test case when slot is not free

**Review by napster1991 — CHANGES_REQUESTED** [review]

File: Parking-Zone/Areas/User/Controllers/ReservationController.cs [hunk 12]
```diff
@@ -41,5 +51,49 @@ public IActionResult FreeSlots(FreeSlotsVM freeSlotsVM)
...
+
+                _reservationService.Insert(reservation);
+
+                ViewBag.SuccessMessage = "Reservation created successfully.";
+
+                return View(reserveVM);
```
Comment (napster1991) [inline Parking-Zone/Areas/User/Controllers/ReservationController.cs]: No need in this return.

**Review by napster1991 — CHANGES_REQUESTED** [review]

File: Parking-Zone/Services/ReservationService.cs [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public class ReservationService : Service<Reservation>, IReservationService
+    {
+        public ReservationService(IReservationRepository repository) : base(repository)
+        {
+        }
+
+        public override void Insert(Reservation entity)
+        {
+            entity.Id = Guid.NewGuid();
```
Comment (napster1991) [inline Parking-Zone/Services/ReservationService.cs:14]: Move this line to base Service class (do the same for other inherited services as well).

**Review by napster1991 — CHANGES_REQUESTED** [review]

File: Parking-Zone/Areas/User/Controllers/ReservationController.cs [hunk 14]
```diff
@@ -41,5 +49,48 @@ public IActionResult FreeSlots(FreeSlotsVM freeSlotsVM)
...
+                return NotFound();
+
+            reserveVM.SlotNumber = slot.Number;
+            reserveVM.ZoneAddress = slot.ParkingZone.Address;
+            reserveVM.ZoneName = slot.ParkingZone.Name;
+
+            if (_slotService.IsSlotFree(slot, DateTime.Parse(reserveVM.StartTime), reserveVM.Duration))
+            {
+                var reservation = reserveVM.MapToModel();
+
```
Comment (napster1991) [inline Parking-Zone/Areas/User/Controllers/ReservationController.cs:80]: Remove unnecessary empty lines

**Review by napster1991 — APPROVED** [review]

**Pull request merged**